components/script/dom/offscreencanvasrenderingcontext2d.rs

```diff
@@ -0,0 +1,951 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use canvas_traits::canvas::{Canvas2dMsg, CanvasMsg};
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasLineJoin;
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasLineCap;
+use dom::bindings::cell::DomRefCell;
+use dom::offscreencanvas::OffscreenCanvas;
+use net_traits::image_cache::CanRequestImages;
+use cssparser::Color as CSSColor;
+use std::cell::Cell;
+use net_traits::image_cache::ImageState;
+use net_traits::image_cache::ImageOrMetadataAvailable;
+use dom::bindings::str::DOMString;
+use net_traits::image_cache::UsePlaceholder;
+use dom::bindings::codegen::UnionTypes::StringOrCanvasGradientOrCanvasPattern;
+use std::{cmp, fmt, mem};
+use dom::imagedata::ImageData;
+use dom::bindings::num::Finite;
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasImageSource;
+use dom::bindings::error::{Error, ErrorResult, Fallible};
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasFillRule;
+use dom::canvasgradient::{CanvasGradient, CanvasGradientStyle};
+use dom::canvaspattern::CanvasPattern;
+use net_traits::image_cache::ImageResponse;
+use net_traits::image::base::PixelFormat;
+use canvas_traits::canvas::{CompositionOrBlending, FillOrStrokeStyle, FillRule};
+use canvas_traits::canvas::{LineCapStyle, LineJoinStyle, LinearGradientStyle};
+use canvas_traits::canvas::{RadialGradientStyle, RepetitionStyle, byte_swap_and_premultiply};
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding;
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding::OffscreenCanvasRenderingContext2DMethods;
+use dom_struct::dom_struct;
+use cssparser::{Parser, ParserInput, RGBA};
+use dom::bindings::reflector::{DomObject, Reflector, reflect_dom_object};
+use ipc_channel::ipc::{self, IpcSender};
+use servo_url::ServoUrl;
+use net_traits::image_cache::ImageCache;
+use euclid::{Transform2D, Point2D, Rect, Size2D};
+use script_traits::ScriptMsg;
+use dom::bindings::root::{Dom, DomRoot, LayoutDom};
+use dom::bindings::codegen::Bindings::ImageDataBinding::ImageDataMethods;
+use std::sync::Arc;
+use std::str::FromStr;
+use dom::globalscope::GlobalScope;
+use dom::node::{document_from_node, window_from_node};
+
+const DEFAULT_WIDTH: u32 = 300;
+const DEFAULT_HEIGHT: u32 = 150;
+
+
+
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[allow(dead_code)]
+enum CanvasFillOrStrokeStyle {
+    Color(RGBA),
+    Gradient(Dom<CanvasGradient>),
+    Pattern(Dom<CanvasPattern>),
+}
+
+#[dom_struct]
+pub struct OffscreenCanvasRenderingContext2D {
```

Comment:
Rather than duplicating so much code in this file, we should be looking at ways to share as much code as possible. Consider extracting common data members and methods into a separate CommonCanvasRenderingContext2d structure that is used as a member by both OffscreenCanvasRenderingContext2D and CanvasRenderingContext2D.

Reply:
Yes @jdm , I agree with you. This was done when we were trying it and I will change this now.
Can you suggest some example files that use the same path of sharing files ? 
It would be convenient to go through an example before trying my hands on this.

Reply:
One example is components/script/dom/webgl2renderingcontext.rs which shares a lot of code with components/script/dom/webglrenderingcontext.rs. components/script/dom/textcontrol.rs is an example of sharing code between components/script/dom/htmlinputelement.rs and components/script/dom/htmltextareaelement.rs.